File: backend/services/bill_chronology.py
from __future__ import annotations

import re
from datetime import datetime, timezone


MONTH_MAP = {
  "jan": 1,
  "feb": 2,
  "mar": 3,
  "apr": 4,
  "may": 5,
  "jun": 6,
  "jul": 7,
  "aug": 8,
  "sep": 9,
  "sept": 9,
  "oct": 10,
  "nov": 11,
  "dec": 12,
}


def get_bill_chronology_timestamp(record: dict) -> float:
  raw_bill_month = str(record.get("bill_month") or "").strip()
  if raw_bill_month:
    text_match = re.search(
      r"\b(jan|feb|mar|apr|may|jun|jul|aug|sep|sept|oct|nov|dec)[a-z]*\b(?:\s+|-|\/)?(20\d{2})?",
      raw_bill_month,
      re.IGNORECASE,
    )
    if text_match:
      month = MONTH_MAP[text_match.group(1).lower()]
      year = int(text_match.group(2) or datetime.now(timezone.utc).year)
      return datetime(year, month, 1, tzinfo=timezone.utc).timestamp()

    numeric_match = re.match(r"^(\d{1,2})[\/ -](20\d{2})$", raw_bill_month)
    if numeric_match:
      month = int(numeric_match.group(1))
      year = int(numeric_match.group(2))
      return datetime(year, month, 1, tzinfo=timezone.utc).timestamp()

  created_at = str(record.get("created_at") or "").strip()
  if created_at:
    normalized = created_at.replace("Z", "+00:00")
    try:
      return datetime.fromisoformat(normalized).timestamp()
    except ValueError:
      pass

  return 0.0


def sort_bills_chronologically(history: list[dict]) -> list[dict]:
  return sorted(history, key=get_bill_chronology_timestamp)

File: lib/bill-chronology.ts
export type BillChronologyRecord = {
  bill_month?: string | null;
  created_at?: string | null;
};

const monthMap: Record<string, number> = {
  jan: 0,
  feb: 1,
  mar: 2,
  apr: 3,
  may: 4,
  jun: 5,
  jul: 6,
  aug: 7,
  sep: 8,
  sept: 8,
  oct: 9,
  nov: 10,
  dec: 11,
};

export function getBillChronologyTime(value: BillChronologyRecord) {
  const raw = value.bill_month?.trim();
  if (raw) {
    const textMatch = raw.match(/\b(jan|feb|mar|apr|may|jun|jul|aug|sep|sept|oct|nov|dec)[a-z]*\b(?:\s+|-|\/)?(20\d{2})?/i);
    if (textMatch) {
      const month = monthMap[textMatch[1].toLowerCase()];
      const year = Number(textMatch[2] ?? new Date().getUTCFullYear());
      return Date.UTC(year, month, 1);
    }

    const numericMatch = raw.match(/^(\d{1,2})[\/ -](20\d{2})$/);
    if (numericMatch) {
      return Date.UTC(Number(numericMatch[2]), Number(numericMatch[1]) - 1, 1);
    }
  }

  const createdAt = value.created_at ? Date.parse(value.created_at) : Number.NaN;
  return Number.isNaN(createdAt) ? 0 : createdAt;
}

export function sortBillsChronologically<T extends BillChronologyRecord>(bills: T[]) {
  return [...bills].sort((left, right) => getBillChronologyTime(left) - getBillChronologyTime(right));
}

export function sortBillsReverseChronologically<T extends BillChronologyRecord>(bills: T[]) {
  return [...bills].sort((left, right) => getBillChronologyTime(right) - getBillChronologyTime(left));
}
